### PR TITLE
Optional Refactoring Examples

### PR DESCRIPTION
**Pull request by jasondegraw**
Pull request overview
---------------------
This PR demonstrates the optional refactoring work with examples in a few files that show the basic approach that we're using and the various outcomes for several of the easy cases.

Some background: modern Fortran has a feature that allows function/subroutine arguments to be designated as "optional". An optional argument cannot be used unless it has been checked as being `present`, which allows for the function or subroutine to modify it's behavior based on the arguments that have been passed. `ObjexxFCL::Optional` objects emulate this feature, but necessarily do some things at runtime (in C++) that could be done at compile time (in Fortran). FORTRAN77, the One True Programming Language, does not support optional arguments. C++ has `std::optional` and default argument values, but `std::optional` serves a slightly different purpose and default argument values do not provide a way to determine if the argument is defaulted.

This PR demonstrates the following cases:

  * The `Optional` arguments are unused and can be removed. The easiest case that appears to be the least frequent. These situations are easy to fix and verify: remove the arguments and recompile, if the arguments are used anywhere there will be compile errors.

  * `Optional` arguments are used to set default values, the presence of the arguments does not change the operation of the function. These situations are best handled with by  requiring the argument or using a default value.

  * `Optional` arguments determine the operation of the function but are limited enough in scope that the function can be split into two (one with the arguments and one without). Care must be taken if `Optional` objects are ever passed as the argument (e.g., if a function taking an optional argument then calls another function and passes the optional as an argument).

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/VariableSpeedCoils.cc [hunk 1]
```diff
@@ -185,12 +183,6 @@ namespace VariableSpeedCoils {
             }
         }
 
-        if (present(OnOffAirFlowRat)) {
```
Comment (jasondegraw) [inline src/EnergyPlus/VariableSpeedCoils.cc:188]: This is just setting up a default value that's a constant, so use a default value instead and eliminate the `Optional`

File: src/EnergyPlus/WaterThermalTanks.cc [hunk 2]
```diff
@@ -10693,14 +10692,7 @@ void WaterThermalTankData::SizeSupplySidePlantConnections(EnergyPlusData &state,
     Real64 tmpUseDesignVolFlowRate = this->UseDesignVolFlowRate;
     Real64 tmpSourceDesignVolFlowRate = this->SourceDesignVolFlowRate;
 
-    int tmpLoopNum;
-    if (!present(LoopNum)) {
-        tmpLoopNum = this->SrcSidePlantLoc.loopNum;
```
Comment (jasondegraw) [inline src/EnergyPlus/WaterThermalTanks.cc:10885]: This time ,the default value isn't a constant, but it is part of the object, so it isn't a big deal to make the argument required and edit the calling locations.

File: src/EnergyPlus/WindowEquivalentLayer.cc [hunk 3]
```diff
@@ -6991,61 +6991,79 @@ bool RB_LWP(CFSLAYER const &L, // RB layer
 }
 
 bool RB_SWP(EnergyPlusData &state,
-            CFSLAYER const &L,                      // RB layer
-            CFSSWP &LSWP,                           // returned: equivalent layer properties set
-            ObjexxFCL::Optional<Real64 const> THETA // incident angle, 0 <= theta <= PI/2
+            CFSLAYER const &L, // RB layer
+            CFSSWP &LSWP,      // returned: equivalent layer properties set
+            const Real64 THETA // incident angle, 0 <= theta <= PI/2
```
Comment (jasondegraw) [inline src/EnergyPlus/WindowEquivalentLayer.cc:6996]: This function uses the presence of the optional to determine which calculation is to be done, but the function isn't called with an `Optional` argument and the function s fairly simple, so break it into two and overload. This is the version with the argument.

File: src/EnergyPlus/WindowEquivalentLayer.cc [hunk 4]
```diff
@@ -6991,61 +6991,79 @@ bool RB_LWP(CFSLAYER const &L, // RB layer
...
+
+    RB_BEAM(state, THETA, RHOBF_BT0, TAUBF_BT0, L.SWP_MAT.TAUSBBB, LSWP.RHOSBBD, LSWP.TAUSBBB, LSWP.TAUSBBD);
+
+    return true;
+}
+
+bool RB_SWP(EnergyPlusData &state,
+            CFSLAYER const &L, // RB layer
+            CFSSWP &LSWP       // returned: equivalent layer properties set
+)
```
Comment (jasondegraw) [inline src/EnergyPlus/WindowEquivalentLayer.cc:7032]: Here's the version of `RB_SWP` that does not have the argument.

File: src/EnergyPlus/WindowEquivalentLayer.cc [hunk 5]
```diff
@@ -7079,62 +7097,79 @@ bool IS_LWP(CFSLAYER const &L, // IS layer
 }
 
 bool IS_SWP(EnergyPlusData &state,
```
Comment (jasondegraw) [inline src/EnergyPlus/WindowEquivalentLayer.cc:7098]: `IS_SWP` is essentially the same case as `RB_SWP`.

File: src/EnergyPlus/WindowEquivalentLayer.cc [hunk 6]
```diff
@@ -7199,86 +7234,95 @@ bool PD_LWP(EnergyPlusData &state,
 }
 
 bool PD_SWP(EnergyPlusData &state,
```
Comment (jasondegraw) [inline src/EnergyPlus/WindowEquivalentLayer.cc:7234]: `PD_SWP` is essentially the same case as `RB_SWP`.

File: src/EnergyPlus/WindowEquivalentLayer.cc [hunk 7]
```diff
@@ -7324,88 +7368,97 @@ bool VB_LWP(EnergyPlusData &state,
 }
 
 bool VB_SWP(EnergyPlusData &state,
```
Comment (jasondegraw) [inline src/EnergyPlus/WindowEquivalentLayer.cc:7368]: `VB_SWP` is essentially the same case as `RB_SWP`.

File: src/EnergyPlus/WindowEquivalentLayer.cc [hunk 8]
```diff
@@ -7675,18 +7728,16 @@ bool IsVBLayer(CFSLAYER const &L)
 }
 
 void BuildGap(EnergyPlusData &state,
-              CFSGAP &G,                               // returned
-              int const GType,                         // gap type (gtyOPENin, gtyOPENout or gtySEALED)
-              Real64 &TAS,                             // gap thickness, m
-              ObjexxFCL::Optional<Real64 const> xTMan, // re density calc -- temp (C) and pressure (Pa)
-              ObjexxFCL::Optional<Real64 const> xPMan  // re density calc -- temp (C) and pressure (Pa)
```
Comment (jasondegraw) [inline src/EnergyPlus/WindowEquivalentLayer.cc:7682]: These two `Optional`s are never used, so just eliminate them.

File: src/EnergyPlus/WindowEquivalentLayer.cc [hunk 9]
```diff
@@ -8039,11 +8082,8 @@ bool FEQX(Real64 const a, // values to compare, fractional tolerance
     Real64 d;
     Real64 tolAbsX;
 
-    if (present(tolAbs)) {
```
Comment (jasondegraw) [inline src/EnergyPlus/WindowEquivalentLayer.cc:8042]: Another case where the optional is being used to set a default value, so use the C++ feature.

File: src/EnergyPlus/VariableSpeedCoils.cc [hunk 10]
```diff
@@ -49,7 +49,6 @@
 
 // ObjexxFCL Headers
 #include <ObjexxFCL/Array.functions.hh>
-#include <ObjexxFCL/Optional.hh>
```
Comment (Myoldmopar) [inline src/EnergyPlus/VariableSpeedCoils.cc:52]: Woo!  Someone needs to fill in the blank  "Every time we remove an ObjexxFCL::Optional, a(n) _________________"

File: src/EnergyPlus/VariableSpeedCoils.cc (hunk 1 above)
Comment (Myoldmopar) [reply src/EnergyPlus/VariableSpeedCoils.cc:188]: Awesome.

File: src/EnergyPlus/VariableSpeedCoils.hh [hunk 11]
```diff
@@ -353,11 +352,11 @@ namespace VariableSpeedCoils {
...
-                               int const SpeedNum,                                   // compressor speed number
-                               Real64 const SpeedRatio,                              // compressor speed ratio
-                               Real64 const SensLoad,                                // Sensible demand load [W]
-                               Real64 const LatentLoad,                              // Latent demand load [W]
-                               ObjexxFCL::Optional<Real64 const> OnOffAirFlowRat = _ // ratio of comp on to comp off air flow rate
+                               int const SpeedNum,                  // compressor speed number
+                               Real64 const SpeedRatio,             // compressor speed ratio
+                               Real64 const SensLoad,               // Sensible demand load [W]
+                               Real64 const LatentLoad,             // Latent demand load [W]
+                               const Real64 OnOffAirFlowRatio = 1.0 // ratio of comp on to comp off air flow rate
```
Comment (Myoldmopar) [inline src/EnergyPlus/VariableSpeedCoils.hh:359]: Awesome.

File: src/EnergyPlus/WaterThermalTanks.cc (hunk 2 above)
Comment (Myoldmopar) [reply src/EnergyPlus/WaterThermalTanks.cc:10885]: :+1:

File: src/EnergyPlus/WindowEquivalentLayer.cc (hunk 8 above)
Comment (Myoldmopar) [reply src/EnergyPlus/WindowEquivalentLayer.cc:7682]: Even better.

**Review by Myoldmopar — APPROVED** [review]
This is awesome.  So what's the plan for this PR specifically?  Are you wanting to just chip away at usages for a while?  Or is this done, and you'll do more in following branches?  Either way you should count the number of Optional usages and each time just post NNN/MMM remaining.  :smile_cat: 

Marking it approved because everything in here looks good to go.  If you are adding more, then just ignore this review.

**Comment by amirroth** [comment]
> This is awesome. So what's the plan for this PR specifically? Are you wanting to just chip away at usages for a while? Or is this done, and you'll do more in following branches? Either way you should count the number of Optional usages and each time just post NNN/MMM remaining. 😸

Just getting rid of `SetDXCoolingCoilData` will reduce the number of Optionals by 5%. I know that sounds like one of my jokes, but grep for yourself and do the math. FWIW I don't mean keep the function and replace the optionals with something else, I mean get rid of it entirely. The entire concept behind a function like this is flawed.

**Comment by jasondegraw** [comment]
@Myoldmopar It may turn into a chip-chip-chip situation later, but I wanted to get the various cases in front of the team so when a bigger PR hits with lots of these it'll make more sense. These are examples of the easier cases, so we'll probably try to knock these out in a PR or two (or three) and then deal with the more involved cases later (cough cough SetDXCoolingCoilData cough cough). I think that'll minimize disruption, but open to suggestions there. My script says there are 61 headers with 153 functions and 394 optionals left to go. 22 of those are in SetDXCoolingCoilData.

I will pull in develop and try to get the formatting straight. I have installed clang-format version 10 and will try to get that working. I've needed to do that for a while.

**Comment by amirroth** [comment]
27 of them are in CurveManager and they will go away as soon as @Myoldmopar's current branch merges.

**Comment by jasondegraw** [comment]
@Myoldmopar I think this is ready to go now, this time without the repeated attempts to fix the format. Yay, me.

**Comment by Myoldmopar** [comment]
All good, thanks @jasondegraw, this will be a great change getting propagated through the codebase.

**Pull request merged**